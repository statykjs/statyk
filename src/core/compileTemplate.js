import path from "node:path";
import fs from "node:fs";
import { parse } from "node-html-parser";
import regexMatchAll from "../utils/regexMatchAll";
import runExpression from "./runExpression";
import JSON from "json-normalize";
import { merge } from "lodash-es";
import instanceComponentScript, {
  scriptCache,
} from "./instanceComponentScript";

export const MUSTACHE_REGEX = /\\?\{\{(.+?)\}\}/gs;

/**
 * evaluate mustache in html template
 * @param {string} html
 * @param {Record<string, string>} attrs
 * @returns {{ html: string, includes: HTMLElement[] }}
 */
function evaluateMustaches(html, attrs) {
  regexMatchAll(MUSTACHE_REGEX, html, (match) => {
    const expression = match[1];
    html = html.replace(match[0], runExpression(expression, attrs));
  });

  return html;
}

/**
 * evaluate mustache in html template
 * @param {string} html
 * @param {string} children
 * @returns {string}
 */
export function evaluateSlots(html, children) {
  const root = parse(html);
  const slots = root.querySelectorAll("slot");
  slots.forEach((slot) => {
    if (children === "" && slot.innerHTML !== "") return;
    slot.replaceWith(children);
  });
  return root.innerHTML;
}

/**
 * append component instance scripts
 * @param {HTMLElement} root
 */
function appendScripts(root) {
  Object.keys(scriptCache).forEach((scriptKey) => {
    if (!scriptCache[scriptKey]) return;
    const body = root.querySelector("body");
    // only append to body, if the script is not already appended
    if (!body || scriptCache[scriptKey].appended) return;

    const script = scriptCache[scriptKey].content;
    const instances = scriptCache[scriptKey].instances.join("\n");
    body.insertAdjacentHTML(
      "afterend",
      `<script>${script}${instances}</script>`
    );
    scriptCache[scriptKey].appended = true;
    scriptCache[scriptKey].el.remove();
  });
}

/**
 * @param {import("node-html-parser/dist/nodes/html").Attributes} element
 * @param {Record<string, any>} vars
 */
function evaluateMustachesInProps(attributes, vars) {
  let props = [];
  Object.keys(attributes).map((attr) => {
    regexMatchAll(MUSTACHE_REGEX, attributes[attr], (match) => {
      const value = JSON.normalizeSync(runExpression(match[1], vars));
      props.push({ attr, value });
    });
  });
  return props;
}

const resolveIncludes = (html, baseFolder) => {
  const root = parse(html);

  const includes = root.querySelectorAll("include[src]");

  includes.forEach((include) => {
    const url = include.getAttribute("src");
    const parsedUrl = path.resolve(baseFolder, url);
    const html = fs.readFileSync(parsedUrl, { encoding: "utf-8" });

    const finalHtml = resolveIncludes(html, baseFolder);

    include.innerHTML = parse(evaluateSlots(finalHtml, include.innerHTML));
  });

  return root.innerHTML;
};

const findIncludesInMustaches = (html) => {
  let includes = [];
  regexMatchAll(MUSTACHE_REGEX, html, (match) => {
    const expression = match[1];
    const parsedHtml = parse(expression);
    includes.push(...parsedHtml.querySelectorAll("include[src]"));
  });
  return includes;
};

/**
 * @param {string} html
 * @param {string} baseFolder
 * @returns
 */
const compileTemplate = (html, baseFolder, vars = {}) => {
  const root = parse(html);
  const includes = root.querySelectorAll("include[src]");

  // 1.  get all includes and evaluate them but not which are in expression
  // 2.  if found mustache -> evaluate mustaches -> recursive compile
  // 3.  else instanceComponentScript -> evaluateMustaches -> recursive compile

  const mustacheIncludes = findIncludesInMustaches(html);
  includes.forEach((include) => {
    const foundMustacheIncludes = mustacheIncludes.filter((el) => {
      return el.toString() === include.toString();
    });
    if (foundMustacheIncludes.length > 0) {
      root.innerHTML = evaluateMustaches(root.innerHTML, vars);
      root.innerHTML = compileTemplate(
        root.innerHTML,
        baseFolder,
        vars
      ).innerHTML;
      return;
    }

    const evaluatedProps = evaluateMustachesInProps(include.attributes, vars);
    evaluatedProps.forEach((prop) => {
      include.setAttribute(prop.attr, prop.value);
    });
    const url = include.getAttribute("src");
    const parsedUrl = path.resolve(baseFolder, url);
    const shtml = fs.readFileSync(parsedUrl, { encoding: "utf-8" });
    const mergedProps = merge(vars, include.attributes);

    include.replaceWith(
      parse(
        evaluateMustaches(
          compileTemplate(
            instanceComponentScript(
              evaluateSlots(shtml, include.innerHTML),
              parsedUrl,
              mergedProps
            ),
            baseFolder,
            mergedProps
          ).innerHTML,
          include.innerHTML
        )
      )
    );
  });

  const finalHtml = parse(evaluateMustaches(root.innerHTML, vars));
  appendScripts(finalHtml);

  return finalHtml;
};

export default compileTemplate;
