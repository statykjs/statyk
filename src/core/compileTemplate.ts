// @ts-check
import fs from "node:fs";
import path from "node:path";
import { HTMLElement, Node, parse } from "node-html-parser";
import merge from "lodash/merge";
// @ts-ignore
import nJSON from "json-normalize";

import { coreRuntime } from "./compile";
import runExpression from "./runExpression";
import regexMatchAll from "../utils/regexMatchAll";
import instanceComponentScript from "./instanceComponentScript";
import { Attributes } from "node-html-parser/dist/nodes/html";
import runCode from "./runCode";
import fm from "../libs/frontmatter";
import isEmpty from "lodash/isEmpty";

export const MUSTACHE_REGEX = /\\?\{\{(.+?)\}\}/gs;

/**
 * evaluate mustache in html template
 */
function evaluateMustaches(
  html: string,
  props: Record<string, string>,
  globalVars: Record<string, any>
) {
  regexMatchAll(MUSTACHE_REGEX, html, (match) => {
    const expression = match[1];
    html = html.replace(match[0], runExpression(expression, props, globalVars));
  });

  return html;
}

/**
 * evaluate mustache in html template
 */
export function evaluateSlots(html: string, children: string | Node) {
  const root = parse(html);
  const slots = root.querySelectorAll("slot");
  slots.forEach((slot) => {
    if (children === "" && slot.innerHTML !== "") return;
    slot.replaceWith(children);
  });
  return root.innerHTML;
}

/**
 * append component instance scripts
 */
function appendScripts(root: HTMLElement) {
  const scriptCache = coreRuntime.caches.scripts;

  scriptCache.keys().forEach((scriptKey) => {
    const value = scriptCache.get(scriptKey);
    if (!value) return;

    // only append to body, if the script is not already appended
    const body = root.querySelector("body");
    if (!body || value.appended) return;

    const script = value.content;
    const instances = value.instances.join("\n");
    body.insertAdjacentHTML(
      "afterend",
      `<script>${script}${instances}</script>`
    );

    value!.el!.remove();
    scriptCache.put(scriptKey, {
      ...value,
      appended: true,
      el: null,
    });
  });
}

function evaluateMustachesInProps(
  attributes: Attributes,
  props: Record<string, any>,
  globalVars: Record<string, any>
) {
  let _props: { attr: string; value: string }[] = [];
  Object.keys(attributes).map((attr) => {
    regexMatchAll(MUSTACHE_REGEX, attributes[attr], (match) => {
      const value = nJSON.normalizeSync(
        runExpression(match[1], props, globalVars)
      );
      _props.push({ attr, value });
    });
  });
  return _props;
}

/**
 * Not in use
 */
const resolveIncludes = (html: string, baseFolder: string) => {
  const root = parse(html);

  const includes = root.querySelectorAll("include[src]");

  includes.forEach((include) => {
    const url = include.getAttribute("src");
    const parsedUrl = path.resolve(baseFolder, url!);
    const html = fs.readFileSync(parsedUrl, { encoding: "utf-8" });

    const finalHtml = resolveIncludes(html, baseFolder);

    include.innerHTML = evaluateSlots(finalHtml, include.innerHTML);
  });

  return root.innerHTML;
};

/**
 * finds and returns all <include /> elements inside mustaches
 */
const findIncludesInMustaches = (html: string): HTMLElement[] => {
  let includes: HTMLElement[] = [];
  regexMatchAll(MUSTACHE_REGEX, html, (match) => {
    const expression = match[1];
    const parsedHtml = parse(expression);
    includes.push(...parsedHtml.querySelectorAll("include[src]"));
  });
  return includes;
};

type FmExtractorReturn = {
  isCodeblock: boolean;
  body: string;
  attributes: Record<string, any>;
  code: string | null;
};
function extractFrontmatter(content: string): FmExtractorReturn {
  try {
    const matter = fm(content);
    if (isEmpty(matter.attributes)) {
      return {
        isCodeblock: true,
        body: matter.body,
        attributes: {},
        code: matter.frontmatter || null,
      };
    }
    return {
      isCodeblock: false,
      body: matter.body,
      attributes: matter.attributes as Record<string, any>,
      code: null,
    };
  } catch {}
  return {
    isCodeblock: false,
    body: content,
    attributes: {},
    code: null,
  };
}

const compileTemplate = async (
  html: string,
  baseFolder: string,
  context: Record<string, any> = {},
  globalVars: Record<string, any> = {}
) => {
  const { body, attributes, code, isCodeblock } = extractFrontmatter(html);
  const root = parse(body);
  const includes = root.querySelectorAll("include[src]");
  let codeResult = {};
  if (isCodeblock) {
    codeResult = await runCode(code!, context);
  }
  codeResult = merge(codeResult, attributes);

  // 1.  get all includes and evaluate them but not which are in expression
  // 2.  if found mustache -> evaluate mustaches -> recursive compile
  // 3.  else instanceComponentScript -> evaluateMustaches -> recursive compile

  const mustacheIncludes = findIncludesInMustaches(body);
  for (const include of includes) {
    const foundMustacheIncludes = mustacheIncludes.filter((el) => {
      return el.toString() === include.toString();
    });

    if (foundMustacheIncludes.length > 0) {
      root.innerHTML = evaluateMustaches(root.innerHTML, context, globalVars);
      root.innerHTML = (await compileTemplate(
        root.innerHTML,
        baseFolder,
        context,
        globalVars
      ))!.innerHTML;
      continue;
    }

    const evaluatedProps = evaluateMustachesInProps(
      include.attributes,
      context,
      codeResult
    );
    evaluatedProps.forEach((prop) => {
      include.setAttribute(prop.attr, prop.value);
    });
    const url = include.getAttribute("src");
    const parsedUrl = path.resolve(baseFolder, url!);
    const shtml = fs.readFileSync(parsedUrl, { encoding: "utf-8" });

    const mergedProps = merge(context, include.attributes);

    const awaited = await compileTemplate(
      instanceComponentScript(
        evaluateSlots(shtml, include.innerHTML),
        parsedUrl,
        mergedProps
      ),
      baseFolder,
      mergedProps,
      codeResult
    );
    include.replaceWith(
      parse(evaluateMustaches(awaited!.innerHTML, mergedProps, codeResult))
    );
  }

  const finalHtml = parse(
    evaluateMustaches(root.innerHTML, context, merge(codeResult, globalVars))
  );
  appendScripts(finalHtml);

  return finalHtml;
};

export default compileTemplate;
