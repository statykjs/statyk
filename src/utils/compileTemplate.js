import path from "node:path";
import fs from "node:fs";
import { parse } from "node-html-parser";
import regexMatchAll from "./regexMatchAll";
import runExpression from "./runExpression";
import JSON from "json-normalize";

export const MUSTACHE_REGEX = /\\?\{\{(.+?)\}\}/gs;

/**
 * evaluate mustache in html template
 * @param {string} html
 * @param {Record<string, string>} attrs
 * @returns {string}
 */
function evaluateMustaches(html, attrs) {
  regexMatchAll(MUSTACHE_REGEX, html, (match) => {
    const expression = match[1];
    html = html.replace(match[0], runExpression(expression, attrs));
  });
  return html;
}

/**
 * evaluate mustache in html template
 * @param {string} html
 * @param {string} children
 * @returns {string}
 */
function evaluateSlots(html, children) {
  const root = parse(html);
  const slots = root.querySelectorAll("slot");
  slots.forEach((slot) => {
    slot.replaceWith(children);
  });
  return root.innerHTML;
}

/**
 * @param {import("node-html-parser/dist/nodes/html").Attributes} element
 * @param {Record<string, any>} vars
 */
function evaluateMustachesInProps(attributes, vars) {
  let props = [];
  Object.keys(attributes).map((attr) => {
    regexMatchAll(MUSTACHE_REGEX, attributes[attr], (match) => {
      const value = JSON.normalizeSync(runExpression(match[1], vars));
      props.push({ attr, value });
    });
  });
  return props;
}

/**
 * @param {string} html
 * @param {string} baseFolder
 * @returns
 */
const compileTemplate = (html, baseFolder, vars = {}) => {
  const root = parse(html);
  const includes = root.querySelectorAll("include[src]");

  includes.forEach((include) => {
    const url = include.getAttribute("src");
    const parsedUrl = path.resolve(baseFolder, url);
    const html = fs.readFileSync(parsedUrl, { encoding: "utf-8" });

    const evaluatedProps = evaluateMustachesInProps(include.attributes, vars);
    evaluatedProps.forEach((prop) => {
      include.setAttribute(prop.attr, prop.value);
    });

    const finalHtml = evaluateMustaches(
      compileTemplate(
        evaluateSlots(html, include.innerHTML),
        baseFolder,
        include.attributes
      ),
      include.attributes
    );
    include.replaceWith(parse(finalHtml));
  });

  return parse(evaluateMustaches(root.innerHTML, vars));
};

export default compileTemplate;
