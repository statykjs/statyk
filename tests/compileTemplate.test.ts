import dedent from "dedent";
import compileTemplate from "../src/core/compileTemplate";

jest.mock("shortid", () => {
  let count = 0
  return { generate: jest.fn(() => {
    count++;
    return count;
  }) };
});

describe("compileTemplate", () => {
  it("should render html", async () => {
    const html = `<p>{{props.msg}}</p>`;

    const res = await compileTemplate(html, "./", { msg: "hello" });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with props", async () => {
    const html = `<p>{{props.count}}: <p>{{props.count + 1}}</p>`;

    const res = await compileTemplate(html, "./", { count: 1 });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render global context values", async () => {
    const html = `<p>{{name}}</p>`;

    const res = await compileTemplate(html, "./", {}, { name: "Anurag" });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with frontmatter yml", async () => {
    const html = dedent`
      ---yml
      name: anuraghazra
      foods: 
        - apple
        - mango
      ---
      <p>{{name}}</p>
      <div>{{map(foods, (food) => '<p>'+food+'</p>')}}</div>
    `;

    const res = await compileTemplate(html, "./");
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with frontmatter js code", async () => {
    const html = dedent`
      ---
      const count = 10;
      return { count };
      ---
      <p>{{count + props.add}}</p>
    `;

    const res = await compileTemplate(html, "./", { add: 1 });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with include", async () => {
    const html = dedent`
      <include src="fixtures/test.html" data="{{props.data}}" />
    `;

    const res = await compileTemplate(html, "./tests", { data: 1 });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with nested includes", async () => {
    const html = dedent`
      <include src="fixtures/nested-1.html" root="{{props.msg}}" />
    `;

    const res = await compileTemplate(html, "./tests", { msg: "root" });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with multiple includes", async () => {
    const html = dedent`
      <include src="fixtures/test.html" data="{{parseInt(props.data)+1}}" />
      <include src="fixtures/test.html" data="{{parseInt(props.data)+2}}" />
    `;

    const res = await compileTemplate(html, "./tests", { data: 1 });
    expect(res.innerHTML).toMatchSnapshot();
  });

  test("frontmatter codeblock should take props", async () => {
    const html = dedent`
      ---
      const url: string = 'https://nice.com/' + props.username;
      return { url };
      ---
      <p>{{url}}</p>
    `;

    const res = await compileTemplate(html, "./tests", {
      username: "anuraghazra",
    });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with partials containing frontmatter codeblock", async () => {
    const html = dedent`
    ---yml
    count: 1
    ---
    <include src="fixtures/frontmatter-code.html" username="{{props.username + '/' + count}}" />
    `;

    const res = await compileTemplate(html, "./tests", {
      username: "anuraghazra",
    });
    expect(res.innerHTML).toMatchSnapshot();
  });

  it("should render with markdown", async () => {
    const html = dedent`
      ### Hello {{props.msg}}
    `;

    const res = await compileTemplate(html, "./tests", { msg: "Anurag" });
    expect(res.innerHTML).toMatchSnapshot();
  });

  test("reusable js components should work", async () => {
    const html = dedent`
      <body>
        <include src="fixtures/counter.html" step={{5}} count={{1}} />
        <include src="fixtures/counter.html" step={{10}} count={{2}} />
      </body>
    `;

    const res = await compileTemplate(html, "./tests", {});
    expect(res.innerHTML).toMatchSnapshot();
  });
});
