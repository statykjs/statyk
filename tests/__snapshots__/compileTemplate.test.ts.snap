// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileTemplate should render global context values 1`] = `"<p>Anurag</p>"`;

exports[`compileTemplate should render html 1`] = `"<p>hello</p>"`;

exports[`compileTemplate should render with frontmatter js code 1`] = `"<p>11</p>"`;

exports[`compileTemplate should render with frontmatter yml 1`] = `
"<p>anuraghazra</p>
<div><p>apple</p>
<p>mango</p></div>"
`;

exports[`compileTemplate should render with props 1`] = `"<p>1: </p><p>2</p>"`;
